torch/optim/adam.py

```diff
@@ -278,7 +280,7 @@ def _single_tensor_adam(params: List[Tensor],
 
             if amsgrad:
                 # Maintains the maximum of all 2nd moment running avg. till now
-                torch.maximum(max_exp_avg_sqs[i], exp_avg_sq, out=max_exp_avg_sqs[i])
+                max_exp_avg_sqs[i][:] = torch.maximum(max_exp_avg_sqs[i].clone(), exp_avg_sq)
```

Comment:
This is a very personal preference but I'm not very confident with advanced indexing assignment in general, can this be done with `copy_()` to make sure of the behavior we get?

But we should also avoid the clone in general.
```suggestion
                prev_exp_avg = max_exp_avg_sqs[i].clone() if differentiable else max_exp_avg_sqs[i]
                max_exp_avg_sqs[i].copy_(torch.maximum(prev_exp_avg, exp_avg_sq))
```

Reply:
the problem here is that if we pass the `differentiable` argument, the JIT and other pieces of code will complain and we will make this non-compat since torchscript does not allow arguments with default values.
I wonder if we can store the flag in the thread local storage set with the differentiable context manager for this check?

Reply:
It is only kwarg-only arguments with default values IIRC ?
You can make it positional as done in https://github.com/pytorch/pytorch/blob/1dfcad84aaad9c095b4dbbdafd5c351e9a712d45/torch/optim/sgd.py#L172-L175 to workaround the torchscript bug.